p2p/src/utils/requester.rs

```diff
@@ -0,0 +1,387 @@
+
+            excluded: HashSet::new(),
+
+            rate_limiter,
+            participants: PrioritySet::new(),
+
+            id: 0,
+            requests: HashMap::new(),
+            deadlines: PrioritySet::new(),
+        }
+    }
+
+    /// Indicate which participants can be sent requests.
+    pub fn reconcile(&mut self, participants: &[PublicKey]) {
+        self.participants
+            .reconcile(participants, self.initial.as_millis());
+        self.rate_limiter.shrink_to_fit();
+    }
+
+    /// Skip a participant for future requests.
+    ///
+    /// Participants added to this list will never be removed (even if dropped
+    /// during `reconcile`, in case they are re-added later).
+    pub fn block(&mut self, participant: PublicKey) {
+        self.excluded.insert(participant);
+    }
+
+    /// Ask for a participant to handle a request.
+    ///
+    /// If `shuffle` is true, the order of participants is shuffled before
+    /// a request is made. This is typically used when a request to the preferred
+    /// participant fails.
+    pub fn request(&mut self, shuffle: bool) -> Option<(PublicKey, ID)> {
+        // Prepare participant iterator
+        let participant_iter = if shuffle {
+            let mut participants = self.participants.iter().collect::<Vec<_>>();
+            participants.shuffle(&mut self.runtime);
+            Either::Left(participants.into_iter())
+        } else {
+            Either::Right(self.participants.iter())
+        };
+
+        // Look for a participant that can handle request
+        for (participant, _) in participant_iter {
+            // Check if me
+            if *participant == self.crypto.public_key() {
+                continue;
+            }
+
+            // Check if excluded
+            if self.excluded.contains(participant) {
+                continue;
+            }
+
+            // Check if rate limit is exceeded (and update rate limiter if not)
+            if self.rate_limiter.check_key(participant).is_err() {
+                continue;
+            }
+
+            // Compute ID
+            //
+            // As long as we don't have u64 requests outstanding, this is ok.
+            let id = self.id;
+            self.id = self.id.wrapping_add(1);
+
+            // Record request issuance time
+            let now = self.runtime.current();
+            self.requests.insert(id, (participant.clone(), now));
+            let deadline = now.checked_add(self.timeout).expect("time overflowed");
+            self.deadlines.put(id, deadline);
+            return Some((participant.clone(), id));
+        }
+        None
+    }
+
+    /// Resolve an outstanding request.
+    pub fn resolved(&mut self, id: ID) {
+        // Remove request
+        let Some((participant, start)) = self.cancel(id) else {
+            return;
+        };
+
+        // Get elapsed time
+        let Ok(elapsed) = self.runtime.current().duration_since(start) else {
+            return;
```

Comment:
tiny nit: consider `Ok(Duration::default())` instead of `return;` due to the non-monotonic possibility of `SystemTime`, which takes `0` as the time and updates the participant's performance (rather than avoiding updating the participant's performance)

Reply:
Good catch!